File: modules/angular2/src/core/compiler/pipe_resolver.ts
import {resolveForwardRef, Injectable} from 'angular2/di';
import {Type, isPresent, BaseException, stringify} from 'angular2/src/core/facade/lang';
import {PipeMetadata} from '../metadata/directives';
import {reflector} from 'angular2/src/core/reflection/reflection';

/**
 * Resolve a `Type` for {@link PipeMetadata}.
 *
 * This interface can be overridden by the application developer to create custom behavior.
 *
 * See {@link Compiler}
 */
@Injectable()
export class PipeResolver {
  /**
   * Return {@link PipeMetadata} for a given `Type`.
   */
  resolve(type: Type): PipeMetadata {
    var metas = reflector.annotations(resolveForwardRef(type));
    if (isPresent(metas)) {
      for (var i = 0; i < metas.length; i++) {
        var annotation = metas[i];
        if (annotation instanceof PipeMetadata) {
          return annotation;
        }
      }
    }
    throw new BaseException(`No Pipe decorator found on ${stringify(type)}`);
  }
}

File: lib/src/core/linker/pipe_resolver.dart
library angular2.src.core.linker.pipe_resolver;

import "package:angular2/src/core/di.dart" show resolveForwardRef, Injectable;
import "package:angular2/src/facade/lang.dart" show Type, isPresent, stringify;
import "package:angular2/src/facade/exceptions.dart" show BaseException;
import "package:angular2/src/core/metadata.dart" show PipeMetadata;
import "package:angular2/src/core/reflection/reflection.dart" show reflector;

bool _isPipeMetadata(dynamic type) {
  return type is PipeMetadata;
}

/**
 * Resolve a `Type` for [PipeMetadata].
 *
 * This interface can be overridden by the application developer to create custom behavior.
 *
 * See [Compiler]
 */
@Injectable()
class PipeResolver {
  /**
   * Return [PipeMetadata] for a given `Type`.
   */
  PipeMetadata resolve(Type type) {
    var metas = reflector.annotations(resolveForwardRef(type));
    if (isPresent(metas)) {
      var annotation = metas.firstWhere(_isPipeMetadata, orElse: () => null);
      if (isPresent(annotation)) {
        return annotation;
      }
    }
    throw new BaseException(
        '''No Pipe decorator found on ${ stringify ( type )}''');
  }
}
